designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+  const eslint = new ESLint()
+
+  for await (const result of eslint.executeOnFiles(patterns)) {
+    print(result)
+  }
+  ```
+
+- If you used the returned object with `await` expression, it returns the array that contains all lint results in random order. This way waits for all lint results then returns it.
+
+  ```js
+  const { ESLint } = require("eslint")
+  const eslint = new ESLint()
+
+  const results = await eslint.executeOnFiles(patterns)
+  // Optionally you can sort the results.
+  // results.sort(ESLint.compareResultsByFilePath)
+  print(results)
+  ```
+
+Either way, we can support "linting in parallel", loading configs/plugins with `import()`, and other stuff that needs asynchronous logic.
+
+##### Move the `usedDeprecatedRules` property
+
+The returned object of `CLIEngine#executeOnFiles()` has the `usedDeprecatedRules` property that includes the deprecated rule IDs which the linting used. But the location doesn't fit the use with `for-await-of` statement. Therefore, this RFC moves the `usedDeprecatedRules` property to each lint result.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+for await (const result of eslint.executeOnFiles(patterns)) {
+  console.log(result.usedDeprecatedRules)
+}
+```
+
+As a side-effect, formatters gets the capability to print the used deprecated rules. (!?)
+
+##### Fail-fast on parallel calling
+
+Because this method updates the cache file, if people call this method in parallel then it causes broken. To prevent the broken, it should throw an error if people called this method while the previous call is still running.
+
+##### Abort linting
+
+The iterator has optional [`return()` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return). The `for-of`/`for-await-of` syntax calls the `return()` method automatically if the execution escaped from the loop by `braek`, `return`, or `throw`. In short, the `return()` method will be called when aborted.
+
+Therefore, ESLint aborts linting when the `return()` method is called. This will mean that the `return()` method terminates all workers once we implement parallel linting.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+for await (const result of eslint.executeOnFiles(patterns)) {
+  if (Math.random() < 0.5) {
+    break // abort linting.
+  }
+}
+```
+
+And it throws an error if you reuse the aborted results.
+
+##### Implementation
+
+<details>
+<summary>A rough sketch of the `executeOnFiles()` method.</summary>
+
+A tiny wrapper of `CLIEngine`.
+
+```js
+class ESLint {
+  executeOnFiles(patterns) {
+    let promise
+    try {
+      const report = this.cliEngine.executeOnFiles(patterns)
+      promise = Promise.resolve(report.results)
+    } catch (error) {
+      promise = Promise.reject(error)
+    }
+
+    return {
+      then(onFulfilled, onRejected) {
+        return promise.then(onFulfilled, onRejected)
+      },
```

Comment:
This won't support the `.catch()` like a full Promise would + is some fairly complex code which pokes at the inner workings of both generators and promises rather than using the built in help that comes with the languages.

Why not separate the two into two individual methods like this?

```javascript
const aggregatedExecuteOnFiles = async function (patterns) {
  const report = this.cliEngine.executeOnFiles(patterns)
  return report.results
}
const executeOnFiles = async function* (patterns) {
  yield aggregatedExecuteOnFiles(patterns)
}
```

Right now `executeOnFiles` wraps `aggregatedExecuteOnFiles`, but eventually it will be `aggregatedExecuteOnFiles` that wraps `executeOnFiles` – right?

Then each method only returns a single type of result and the built in syntax helps ensure they are fully valid instances of their each respective type 🙂

Reply:
Because we have to use the inner working from the first (see also https://github.com/eslint/rfcs/pull/40#discussion_r334930763). Additional complexity is small. And the "inner working" is defined clearly in the language spec, we can use it safely.